### PR TITLE
Promises refactor plus bugfix

### PR DESCRIPTION
**Pull request by robmoffat**
Hi Guys,

I've reviewed all the app-channel and user-channel tests against the spec, based on our discovery yesterday that we weren't passing in the context type into `addContextListener` properly.  

Can we review this?   I've run this against my local finsemble - would be nice to know how this works on Glue too.  

This PR builds on top of @ksgeorgieva 's promises-refactor PR.

### CODE REVIEW

**Comment by kriswest** [comment]
FYI refactor-test-app-channels / PR #142 adds one commit to promises-refactor / #138  hence I suggest that:
- we close #138
- retarget this PR #144 at #142's branch (refactor-test-app-channels)
- review and merge this PR
- return to #142 and check off which comments are now resolved by your changes
- continue forward with #142 and try to get that merged to main when comments are dealt with

File: src/test/common/channel-control.ts [hunk 1]
```diff
@@ -19,9 +19,9 @@ export interface ChannelControl<X, Y> {
   
     // listening
     initCompleteListener(testId: string): Promise<Y>
-    setupAndValidateListener1(channel: X, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): void | Promise<void>;
-    setupAndValidateListener2(channel: X, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): void | Promise<void>;
-    setupContextChecker(channel: X, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): Promise<void>;
+    setupAndValidateListener1(channel: X | null, expectedContextType: string | null, errorMessage: string, onComplete: (ctx: Y) => void): void | Promise<void>;
```
Comment (kriswest) [inline src/test/common/channel-control.ts:22]: I think we ought to separate the expectedContextType and the type of the listener added - particularly as I think the expected context type always has to be set.

```suggestion
    setupAndValidateListener1 = (channel: X | null, listenerContextType: string | null, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): void | Promise<void>;
```

File: src/test/common/channel-control.ts [hunk 2]
```diff
@@ -19,9 +19,9 @@ export interface ChannelControl<X, Y> {
...
+    setupAndValidateListener2(channel: X | null, expectedContextType: string | null, errorMessage: string, onComplete: (ctx: Y) => void): void | Promise<void>;
+    setupContextChecker(channel: X, requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): Promise<void>;
```
Comment (kriswest) [inline src/test/common/channel-control.ts:24]: I think one more util function is needed here for cases where context should NOT be received (ensuring we can have a better error message than
> expected someContexttype to equal "unexpected-context"

```suggestion
    setupContextChecker(channel: X, requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): Promise<void>;
```

```suggestion
    validateContextIsNotReceived = (channel: X, listenerContextType: string | null, errorMessage: string, onComplete: (ctx: Y) => void): void | Promise<void>;
    setupContextChecker(channel: X, requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Y) => void): Promise<void>;
```

File: src/test/common/fdc3.app-channels.ts [hunk 3]
```diff
@@ -20,7 +20,7 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
         const testChannel = await cc.createTestChannel()
         const resolveExecutionCompleteListener = cc.initCompleteListener(acTestId)
         let receivedContext = false;
-        await cc.setupAndValidateListener1(testChannel, "fdc3.instrument", errorMessage, () => { receivedContext = true })
+        await cc.setupAndValidateListener1(testChannel, null, errorMessage, () => { receivedContext = true })
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:23]: based on above suggested modification to util function. There are a bunch of other places similar changes will be needed not currently covered in the PR.
```suggestion
        await cc.setupAndValidateListener1(testChannel, null, "fdc3.instrument", errorMessage, () => { receivedContext = true })
```

File: src/test/common/fdc3.app-channels.ts [hunk 4]
```diff
@@ -128,7 +128,7 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
 
         const testChannel = await cc.createTestChannel("a-different-test-channel");
         await cc.setupAndValidateListener1(testChannel, "unexpected-context", errorMessage, () => { /*noop*/ })
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:130]: Based on suggested extra util function
```suggestion
        await cc.validateContextIsNotReceived(testChannel, null, errorMessage, () => { /*noop*/ })
```

File: src/test/common/fdc3.app-channels.ts [hunk 5]
```diff
@@ -128,7 +128,7 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
...
-        await cc.openChannelApp(acTestId7, undefined, APP_CHANNEL_AND_BROADCAST)
+        await cc.openChannelApp(acTestId7, undefined, APP_CHANNEL_AND_BROADCAST_TWICE)
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:131]: I _think_ this change is making it broadcast instrument and contact contexts rather than just instrument. Doesn't really break the test at all, but varies slightly from the description above - but back towards the actual test definition 👍  - I can't suggest a correction to the description above but it should be fixed

File: src/test/common/fdc3.app-channels.ts [hunk 6]
```diff
@@ -137,12 +137,18 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
...
 
-        let testChannel = await cc.createTestChannel()
+        let testChannel1 = await cc.createTestChannel()
+        let receivedContext = false;
+
         const resolveExecutionCompleteListener = cc.initCompleteListener(acTestId8)
-        testChannel = await cc.createTestChannel("a-different-test-channel");
-        await cc.setupAndValidateListener1(testChannel, "unexpected-context", errorMessage, () => { /*noop*/ })
-        await cc.openChannelApp(acTestId8, undefined, APP_CHANNEL_AND_BROADCAST)
+        let testChannel2 = await cc.createTestChannel("a-different-test-channel");
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:144]: Note this clashes with #142 where the argument goes away - personally I prefer this explicit style as the test is more readable. Have suggested changing the function name to `createARandomTestChannel` to fix the semantics

File: src/test/common/fdc3.app-channels.ts [hunk 7]
```diff
@@ -137,12 +137,18 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
...
+        await cc.setupAndValidateListener1(testChannel1, "fdc3.instrument", errorMessage, () => { receivedContext = true })
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:145]: based on suggested change to util function
```suggestion
        await cc.setupAndValidateListener1(testChannel1, "fdc3.instrument", "fdc3.instrument", errorMessage, () => { receivedContext = true })
```

File: src/test/common/fdc3.app-channels.ts [hunk 8]
```diff
@@ -137,12 +137,18 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
...
+        await cc.setupAndValidateListener2(testChannel2, "unexpected-context", errorMessage, () => { /*noop*/ })
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:146]:
```suggestion
        await cc.setupAndValidateListener2(testChannel2, "unexpected-context", errorMessage, () => { /*noop*/ })
        await cc.validateContextIsNotReceived(testChannel2, null, errorMessage, () => { /*noop*/ })
```

File: src/test/common/fdc3.app-channels.ts [hunk 9]
```diff
@@ -137,12 +137,18 @@ export function createAppChannelTests(cc: ChannelControl<any,any>, documentation
...
+        await cc.openChannelApp(acTestId8, undefined, APP_CHANNEL_AND_BROADCAST_TWICE)
```
Comment (kriswest) [inline src/test/common/fdc3.app-channels.ts:147]: Needs to join the correct channel
```suggestion
        await cc.openChannelApp(acTestId8, testChannel1.id, APP_CHANNEL_AND_BROADCAST_TWICE)
```

File: src/test/common/fdc3.user-channels.ts [hunk 10]
```diff
@@ -19,7 +19,7 @@ export function createUserChannelTests(cc: ChannelControl<any,any>, documentatio
 
         const resolveExecutionCompleteListener = cc.initCompleteListener(scTestId1)
         let receivedContext = false;
-        await cc.setupAndValidateListener1(null, "fdc3.instrument", errorMessage, () => receivedContext = true);
+        await cc.setupAndValidateListener1(null, null, errorMessage, () => receivedContext = true);
```
Comment (kriswest) [inline src/test/common/fdc3.user-channels.ts:22]:
```suggestion
        await cc.setupAndValidateListener1(null, "fdc3.instrument", "fdc3.instrument", errorMessage, () => receivedContext = true);
```

File: src/test/common/fdc3.user-channels.ts [hunk 11]
```diff
@@ -37,7 +37,7 @@ export function createUserChannelTests(cc: ChannelControl<any,any>, documentatio
         const resolveExecutionCompleteListener = cc.initCompleteListener(scTestId2)
         const channel = await cc.retrieveAndJoinChannel(1);
         let receivedContext = false;
-        await cc.setupAndValidateListener1(null, "fdc3.instrument", errorMessage, () => receivedContext = true);
+        await cc.setupAndValidateListener1(null, null, errorMessage, () => receivedContext = true);
```
Comment (kriswest) [inline src/test/common/fdc3.user-channels.ts:40]:
```suggestion
        await cc.setupAndValidateListener1(null, "fdc3.instrument", "fdc3.instrument", errorMessage, () => receivedContext = true);
```

File: src/test/common/fdc3.user-channels.ts [hunk 12]
```diff
@@ -56,7 +56,7 @@ export function createUserChannelTests(cc: ChannelControl<any,any>, documentatio
         await cc.openChannelApp(scTestId3, channel.id, JOIN_AND_BROADCAST);
         await cc.joinChannel(channel);
         let receivedContext = false;
-        await cc.setupAndValidateListener1(null, "fdc3.instrument", errorMessage, () => receivedContext = true);
+        await cc.setupAndValidateListener1(null, null, errorMessage, () => receivedContext = true);
```
Comment (kriswest) [inline src/test/common/fdc3.user-channels.ts:59]:
```suggestion
        await cc.setupAndValidateListener1(null, "fdc3.instrument", "fdc3.instrument", errorMessage, () => receivedContext = true);
```

File: src/test/v1.2/advanced/channels-support-1.2.ts [hunk 13]
```diff
@@ -105,13 +105,17 @@ export class ChannelControl1_2 implements ChannelControl<Channel, Context> {
...
     } else {
-      listener1 = fdc3.addContextListener(null, (context) => {
-        expect(context.type).to.be.equals(expectedContextType, errorMessage);
+      listener1 = fdc3.addContextListener(expectedContextType, (context) => {
+        if (expectedContextType != null) {
+          expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        }
         onComplete(context);
       });
     }
```
Comment (kriswest) [inline src/test/v1.2/advanced/channels-support-1.2.ts:121]: This isn't quite right as we're using the same argument for both the listener type and the expected type. We need to be able to listen for null but expect a particular type in many tests...

Due to the deleted lines I can't put a suggestion in here but I think this function should be:

```typescript
setupAndValidateListener1 = (channel: Channel, listenerContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
    if (channel) {
      listener1 = channel.addContextListener(listenerContextType, (context) => {
        expect(context.type).to.be.equals(expectedContextType, errorMessage);
        onComplete(context);
      });
    } else {
      listener1 = fdc3.addContextListener(listenerContextType, (context) => {
        expect(context.type).to.be.equals(expectedContextType, errorMessage);
        onComplete(context);
      });
    }
    validateListenerObject(listener1);
  }
  ```

same applies to the second copy of the function below

File: src/test/v1.2/advanced/channels-support-1.2.ts [hunk 14]
```diff
@@ -121,23 +125,27 @@ export class ChannelControl1_2 implements ChannelControl<Channel, Context> {
 
   setupAndValidateListener2 = (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
```
Comment (kriswest) [inline src/test/v1.2/advanced/channels-support-1.2.ts:126]: see above

File: src/test/v1.2/advanced/channels-support-1.2.ts [hunk 15]
```diff
@@ -121,23 +125,27 @@ export class ChannelControl1_2 implements ChannelControl<Channel, Context> {
...
+      listener2 = fdc3.addContextListener(expectedContextType, (context) => {
+        if (expectedContextType != null) {
+          expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        }
         onComplete(context);
       });
     }
 
     validateListenerObject(listener2);
   }
```
Comment (kriswest) [inline src/test/v1.2/advanced/channels-support-1.2.ts:145]: I suggest an additional util function here for cases where no context is expected to arrive (which we do in many places and currently output a poor error message)

```suggestion

validateContextIsNotReceived = (channel: Channel, listenerContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
    if (channel) {
      listener1 = channel.addContextListener(listenerContextType, (context) => {
        assert.fail(errorMessage);
        onComplete(context);
      });
    } else {
      listener1 = fdc3.addContextListener(listenerContextType, (context) => {
        assert.fail(errorMessage);
        onComplete(context);
      });
    }
    validateListenerObject(listener1);
  }

```

File: src/test/v1.2/advanced/channels-support-1.2.ts [hunk 16]
```diff
@@ -121,23 +125,27 @@ export class ChannelControl1_2 implements ChannelControl<Channel, Context> {
...
 
-  setupContextChecker = async (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
+  setupContextChecker = async (channel: Channel,  requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
     //Retrieve current context from channel
-    const context = await channel.getCurrentContext();
+    const context = (requestedContextType == undefined) ? await channel.getCurrentContext() : await channel.getCurrentContext(requestedContextType);
```
Comment (kriswest) [inline src/test/v1.2/advanced/channels-support-1.2.ts:148]:
```suggestion
    const context = (requestedContextType == null) ? await channel.getCurrentContext() : await channel.getCurrentContext(requestedContextType);
```
`requestedContextType ` is in teh middle of the args list so it will need to be `null` rather than `undefined`.

File: src/test/v2.0/advanced/channels-support-2.0.ts [hunk 17]
```diff
@@ -108,13 +108,17 @@ export class ChannelControl2_0 implements ChannelControl<Channel, Context> {
 
   setupAndValidateListener1 = async (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void) => {
```
Comment (kriswest) [inline src/test/v2.0/advanced/channels-support-2.0.ts:109]: See suggestion on 1.2 version of this function. We *have* to separate the requested context type and expected context type to make this work

File: src/test/v2.0/advanced/channels-support-2.0.ts [hunk 18]
```diff
@@ -124,26 +128,30 @@ export class ChannelControl2_0 implements ChannelControl<Channel, Context> {
 
   setupAndValidateListener2 = async (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void) => {
```
Comment (kriswest) [inline src/test/v2.0/advanced/channels-support-2.0.ts:129]: as above

File: src/test/v2.0/advanced/channels-support-2.0.ts [hunk 19]
```diff
@@ -124,26 +128,30 @@ export class ChannelControl2_0 implements ChannelControl<Channel, Context> {
...
+      listener2 = await fdc3.addContextListener(expectedContextType, (context) => {
+        if (expectedContextType != null) {
+          expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        }
         onComplete(context);
       });
     }
 
     validateListenerObject(listener2);
   }
```
Comment (kriswest) [inline src/test/v2.0/advanced/channels-support-2.0.ts:148]: See suggest additional util function suggested in 1.2 tests

File: src/test/v2.0/advanced/channels-support-2.0.ts [hunk 20]
```diff
@@ -124,26 +128,30 @@ export class ChannelControl2_0 implements ChannelControl<Channel, Context> {
...
 
-  setupContextChecker = async (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void) => {
+  setupContextChecker = async (channel: Channel,  requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
     //Retrieve current context from channel
-    await channel.getCurrentContext().then(async (context) => {
-      expect(context.type).to.be.equals(expectedContextType, errorMessage);
-      onComplete(context);
-    });
+    const context = (requestedContextType == undefined) ? await channel.getCurrentContext() : await channel.getCurrentContext(requestedContextType);
```
Comment (kriswest) [inline src/test/v2.0/advanced/channels-support-2.0.ts:151]: see comments on 1.2 version of function
```suggestion
    const context = (requestedContextType == null) ? await channel.getCurrentContext() : await channel.getCurrentContext(requestedContextType);
```

File: src/test/v2.0/advanced/channels-support-2.0.ts [hunk 21]
```diff
@@ -124,26 +128,30 @@ export class ChannelControl2_0 implements ChannelControl<Channel, Context> {
...
+        }
         onComplete(context);
       });
     }
 
     validateListenerObject(listener2);
   }
 
-  setupContextChecker = async (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void) => {
+  setupContextChecker = async (channel: Channel,  requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
```
Comment (kriswest) [inline src/test/v2.0/advanced/channels-support-2.0.ts:149]:
```suggestion
  setupContextChecker = async (channel: Channel,  requestedContextType: string | null, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
```

File: src/test/v1.2/advanced/channels-support-1.2.ts [hunk 22]
```diff
@@ -121,23 +125,27 @@ export class ChannelControl1_2 implements ChannelControl<Channel, Context> {
...
+        }
         onComplete(context);
       });
     }
 
     validateListenerObject(listener2);
   }
 
-  setupContextChecker = async (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
+  setupContextChecker = async (channel: Channel,  requestedContextType: string, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
```
Comment (kriswest) [inline src/test/v1.2/advanced/channels-support-1.2.ts:146]: See below comment
```suggestion
  setupContextChecker = async (channel: Channel,  requestedContextType: string | null, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): Promise<void> => {
```

**Review by kriswest — CHANGES_REQUESTED** [review]
Heading in the right direction but I think there's a problem being introduced by conflated teh requested context listner type with the expected type - I.e. I think you need to separate these as you are doing in the `setupContextChecker` fn.

Also suggesting an addition util function for when context should not be received. That will improve the failure error message + make the code a bit more readable + ensure that we are not setting up context listeners for an "unexpected-context" (introduced in #138)
 
Also see comments on #142. I think this PR should be re-targeted at that one (to catch the extra commit it adds) and #138 closed. Then when this is merged we can continue on #142 and confirm all the comments are dealt with before merging that

File: src/test/common/fdc3.app-channels.ts (hunk 4 above)
Comment (robmoffat) [reply src/test/common/fdc3.app-channels.ts:130]: done in Kalina's PR

File: src/test/common/fdc3.app-channels.ts (hunk 5 above)
Comment (robmoffat) [reply src/test/common/fdc3.app-channels.ts:131]: yes, this is based on the test specification

File: src/test/v2.0/advanced/channels-support-2.0.ts (hunk 21 above)
Comment (robmoffat) [reply src/test/v2.0/advanced/channels-support-2.0.ts:149]: done

File: src/test/common/channel-control.ts (hunk 1 above)
Comment (kriswest) [reply src/test/common/channel-control.ts:22]: Not true - I found more than one test where a null listener should be added and a specified type received.
